ui/components/src/ui-components/tree/controlled/TreeNodeLoader.ts

```diff
@@ -359,16 +359,24 @@ function updateChildren(
     return;
   }
 
-  model.setChildren(
-    parentId,
-    hierarchyItems.map(({ item }) => convertToTreeModelNodeInput(item)),
-    startIndex,
-  );
+  let offset = startIndex;
+  for (const hierarchyItem of hierarchyItems) {
+    const nodeInput = convertToTreeModelNodeInput(hierarchyItem.item);
+    const existingNode = model.getNode(hierarchyItem.item.id);
+
+    // if same item exists in the same position and is expanded update it without removing it's subtree
+    if (!existingNode || !existingNode.isExpanded || model.getChildOffset(parentId, existingNode.id) !== offset || nodeInput.numChildren === 0) {

```

Comment:
I'm probably missing something obvious, but why just finding existing node is not enough and we're additionally checking for `isExpanded` and offset?

Reply:
My idea was to always overwrite existing node (as it was doing before my changes) instead of updating if it would not affect visible nodes. So the only case when we cannot overwrite node is when node is expanded and there are children visible under the node because it would remove those children from tree model.